executor/cleaner/cleaner_test.go

```diff
@@ -0,0 +1,113 @@
+package cleaner_test
+
+import (
+	"context"
+	"testing"
+
+	k8upv1a1 "github.com/vshn/k8up/api/v1alpha1"
+	"github.com/vshn/k8up/executor/cleaner"
+	"github.com/vshn/k8up/job"
+
+	"github.com/go-logr/logr"
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestCleanup(t *testing.T) {
+	jobs := jobList(3, 2, 2)
+	client := newMockClient(jobs)
+
+	objCleaner := &cleaner.ObjectCleaner{Client: client, Limits: newLimiter(1, 1), Log: logr.DiscardLogger{}}
+	err := objCleaner.CleanOldObjects(context.TODO(), jobs.GetJobObjects())
+	assert.NoError(t, err)
+
+	afterClean := &k8upv1a1.RestoreList{}
+	assert.NoError(t, client.List(context.TODO(), afterClean))
+	runningJobs, failedJobs, successfulJobs := job.GroupJobsByStatus(afterClean.GetJobObjects())
+	assert.Equal(t, 3, len(runningJobs))
+	assert.Equal(t, 1, len(failedJobs))
+	assert.Equal(t, 1, len(successfulJobs))
+}
+
+func newLimiter(maxFailed, maxSuccessful int) limiter {
+	return limiter{maxFailed: &maxFailed, maxSuccessful: &maxSuccessful}
+}
+
+type limiter struct {
+	maxFailed, maxSuccessful *int
+}
+
+func (l limiter) GetSuccessfulJobsHistoryLimit() *int {
+	return l.maxSuccessful
+}
+
+func (l limiter) GetFailedJobsHistoryLimit() *int {
+	return l.maxFailed
+}
+
+func newMockClient(jobs *k8upv1a1.RestoreList) client.Client {
+	objs := make([]client.Object, len(jobs.Items))
+	for i := range jobs.Items {
+		objs[i] = &jobs.Items[i]
+	}
+
+	scheme := runtime.NewScheme()
+	k8upv1a1.AddToScheme(scheme)
+	return fake.NewClientBuilder().WithScheme(scheme).WithObjects(objs...).Build()
+}
```

Comment:
The cleaner is integration tested in the `CheckTestSuite`. 

The problem with integration tests are that they are usually harder to debug. e.g. There must be a correct environment around which can be complicated with editor integrations. And they are slower. I prefer unit tests as long as possible and create more high level integration tests.

Reply:
> correct environment around which can be complicated with editor integrations

We have automated this and thanks to plain testing.T any IDE should be able to integrate nicely. 
Did you look at the env test suite here: https://github.com/vshn/k8up/blob/master/controllers/envsuite_test.go ?
The debugging is now the same as when using the fake client (you get a plain k8s client object).

It just feels kind of wrong to start using the fake client here and become inconsistent with other tests that we have just because it feels "slow"...